### PR TITLE
Fix using structure from insertion tables in case of defaults and explicit insert columns

### LINKED ISSUE #54619: column with DEFAULT condition is not evaluated correctly when importing from Native file
**Describe the unexpected behaviour**

If importing data from a Native file using clickhouse-client older than 23.1 and there is a column with DEFAULT condition, this condition will not be evaluated correctly. 

**How to reproduce**

https://fiddle.clickhouse.com/5a0a8dc6-238b-4b07-b4b9-38ada65f78e1

and repro code here:

```
CREATE TABLE default.generate
(
    `key` String,
    `condition` UInt8,
)
ENGINE = MergeTree()
ORDER BY tuple();

INSERT INTO default.generate VALUES ('a', 0);
INSERT INTO default.generate VALUES ('b', 0);
INSERT INTO default.generate VALUES ('c', 0);

SELECT * FROM default.generate INTO OUTFILE '/var/lib/clickhouse/user_files/generate.native' FORMAT Native;

CREATE TABLE default.repro
(
    `key` String,
    `condition` UInt8,
    `derived_condition` UInt8 DEFAULT (condition = 0),
)
ENGINE = MergeTree()
ORDER BY tuple();

INSERT INTO default.repro FROM INFILE '/var/lib/clickhouse/user_files/generate.native' FORMAT Native;

SELECT * FROM default.repro;

┌─key─┬─condition─┬─derived_condition─┐
│ c   │         0 │                 0 │
│ a   │         0 │                 0 │
│ b   │         0 │                 0 │
└─────┴───────────┴───────────────────┘
```

derived condition should be evaluated as 1 and not 0

* Which ClickHouse server version to use

Older than 23.1.

**Expected behavior**

```
SELECT * FROM default.repro;

┌─key─┬─condition─┬─derived_condition─┐
│ c   │         0 │                 1 │
│ a   │         0 │                 1 │
│ b   │         0 │                 1 │
└─────┴───────────┴───────────────────┘
```

**Additional context**

The bug has been fixed in 23.2 (clickhouse-client) and I'm checking in which commit.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix using structure from insertion tables in case of defaults and explicit insert columns. Closes https://github.com/ClickHouse/ClickHouse/issues/54619

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit b9c28ef1f7f9ae696e13d5ad8cdea7acea12a4ac with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54655/b9c28ef1f7f9ae696e13d5ad8cdea7acea12a4ac/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/02879_use_structure_from_insertion_table_with_defaults.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+$CLICKHOUSE_LOCAL -q "select 42 as x format Native" > $CLICKHOUSE_TEST_UNIQUE_NAME.native
+$CLICKHOUSE_LOCAL -n -q "
+create table test (x UInt64, y UInt64 default 42) engine=Memory;
```
Comment (antonio2368) [inline tests/queries/0_stateless/02879_use_structure_from_insertion_table_with_defaults.sh:9]: 42 is a nice number but I would change the default here so you are sure that y wasn't filled with value from file.

File: src/Interpreters/Context.h [hunk 2]
```diff
@@ -713,9 +719,11 @@ class Context: public std::enable_shared_from_this<Context>
 
     void killCurrentQuery() const;
 
-    bool hasInsertionTable() const { return !insertion_table.empty(); }
-    void setInsertionTable(StorageID db_and_table) { insertion_table = std::move(db_and_table); }
-    const StorageID & getInsertionTable() const { return insertion_table; }
+    bool hasInsertionTable() const { return !insertion_table_info.table.empty(); }
+    bool hasInsertionTableColumnNames() const { return insertion_table_info.column_names.has_value(); }
+    void setInsertionTable(StorageID db_and_table, const std::optional<Names> & column_names = std::nullopt) { insertion_table_info = {std::move(db_and_table), column_names}; }
```
Comment (antonio2368) [inline src/Interpreters/Context.h:724]:
```suggestion
    void setInsertionTable(StorageID db_and_table, std::optional<Names> column_names = std::nullopt) { insertion_table_info = {std::move(db_and_table), std::move(column_names)}; }
```

File: src/Interpreters/InterpreterInsertQuery.cpp [hunk 3]
```diff
@@ -149,6 +149,24 @@ Block InterpreterInsertQuery::getSampleBlock(
...
 }
 
+std::optional<Names> InterpreterInsertQuery::getInsertColumnNames() const
+{
+    auto const * insert_query = query_ptr->as<ASTInsertQuery>();
+    if (!insert_query || !insert_query->columns)
+        return std::nullopt;
+
+    auto table = DatabaseCatalog::instance().getTable(getDatabaseTable(), getContext());
+    Names names;
```
Comment (antonio2368) [inline src/Interpreters/InterpreterInsertQuery.cpp:159]: reserve

Comment (Avogar) [reply src/Interpreters/InterpreterInsertQuery.cpp:159]: Copied similar code from `getSampleBlock` and didn't noticed it, good catch, will add it to both places

File: tests/queries/0_stateless/02879_use_structure_from_insertion_table_with_defaults.sh (hunk 1 above)
Comment (Avogar) [reply tests/queries/0_stateless/02879_use_structure_from_insertion_table_with_defaults.sh:9]: Sure, good idea

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**